cudaaligner/src/hirschberg_myers_gpu.cu

```
@@ -505,80 +483,159 @@ __device__ void hirschberg_myers_single_char_warp(int8_t*& path, int32_t* path_l
+    T* cur_end_;
+    T* buffer_end_;
 };
 
 template <typename T>
-class warp_shared_stack
+class parallel_warp_shared_stack
 {
 public:
-    __device__ warp_shared_stack(char* buffer_begin, char* buffer_end)
-        : buffer_begin_(buffer_begin), cur_end_(buffer_begin), buffer_end_(buffer_end)
+    __device__ parallel_warp_shared_stack(parallel_warp_shared_stack_state<T>* data, T* buffer_begin, T* buffer_end)
+        : data_(data)
     {
-        assert(buffer_begin_ < buffer_end_);
+        assert(buffer_begin < buffer_end);
+        if (threadIdx.x == 0 && threadIdx.y == 0 && threadIdx.z == 0)
+        {
+            data_->mutex_        = 0;
+            data_->active_warps_ = 0;
+            data_->buffer_begin_ = buffer_begin;
+            data_->cur_end_      = buffer_begin;
+            data_->buffer_end_   = buffer_end;
+        }
+        __syncthreads();
+        if (threadIdx.x % warp_size == 0)
+        {
+            atomicAdd(&data_->active_warps_, 1);
+        }
+        __syncthreads();
     }
 
-    __device__ bool inline push(T const& t, unsigned warp_mask = 0xffff'ffff)
+    __device__ bool inline push(T const& t)
     {
-        if (buffer_end_ - cur_end_ >= sizeof(T))
+        lock_mutex_high_priority();
+        bool success = false;
+        __syncwarp();
+
+        if (data_->buffer_end_ - data_->cur_end_ >= 1)
         {
-            __syncwarp(warp_mask);
-            if (threadIdx.x == 0)
+            __syncwarp();
+            if (threadIdx.x % warp_size == 0)
             {
-                std::memcpy(cur_end_, &t, sizeof(T));
+                *(data_->cur_end_) = t;
+                ++(data_->cur_end_);
             }
-            __syncwarp(warp_mask);
-            cur_end_ += sizeof(T);
-            return true;
+            success = true;
         }
         else
         {
-            if (threadIdx.x == 0)
+            if (threadIdx.x % warp_size == 0)
             {
                 printf("ERROR: stack full!");
```

This is from a previous PR, but should this be an assert?

---

No, this should not be checked with an assert. It is something that might actually happen for valid input (if it is too large) and we need to report an error.

---

The error mechanism could need some improvement though...

---

Let's file an issue to improve the error mechanism later? So the failed alignments can be handled on the CPU potentially